yt/yt/server/controller_agent/controllers/multi_job_manager.cpp

```diff
@@ -0,0 +1,198 @@
+#include "multi_job_manager.h"
+#include "job_info.h"
+#include "task.h"
+
+#include <yt/yt/server/controller_agent/controllers/task_host.h>
+#include <yt/yt/server/lib/controller_agent/helpers.h>
+#include <yt/yt/server/lib/controller_agent/progress_counter.h>
+
+#include <yt/yt/ytlib/controller_agent/serialize.h>
+#include <yt/yt/ytlib/scheduler/config.h>
+
+#include <util/generic/algorithm.h>
+
+namespace NYT::NControllerAgent::NControllers {
+
+using namespace NChunkPools;
+
+////////////////////////////////////////////////////////////////////////////////
+
+TMultiJobManager::TMultiJobManager(
+    TTask* task,
+    NLogging::TLogger logger)
+    : JobCounter_(New<TProgressCounter>()), Task_(task), Logger(logger)
+{ }
+
+i64 TMultiJobManager::GetPendingCandidatesDataWeight() const
+{
+    return PendingDataWeight_;
+}
+
+int TMultiJobManager::GetPendingJobCount() const
+{
+    return JobCounter_->GetPending();
+}
+
+int TMultiJobManager::GetTotalJobCount() const
+{
+    return JobCounter_->GetTotal();
+}
+
+std::pair<NChunkPools::IChunkPoolOutput::TCookie, int> TMultiJobManager::PeekJobCandidate() const
+{
+    YT_LOG_WARNING("PeekJobCandidate");
+    YT_VERIFY(!PendingCookies_.empty());
+    auto cookie = *PendingCookies_.begin();
+    auto replicas = GetOrCrash(CookieToReplicas_, cookie);
+    return {cookie, replicas->Secondaries.size() - replicas->Pending + 1};
```

Comment:
Is it correct? If some job in the middle was aborted we should return it. I guess it should be `repliacs->Secondaries[replicas->Secondaries.size() - replicas->Pending]` or something like that.

UPD: probably we will have "restart on job restart" but it seems dangerous anyway.

Reply:
Yes, the plan was that everything would restart on any job failing, so no out of order jobs.